src/blocks/weather/open_weather_map.rs

```diff
@@ -79,24 +94,198 @@ struct ApiWeather {
             .or_else(|| {
                 self.config
                     .coordinates
                     .as_ref()
                     .map(|(lat, lon)| format!("lat={lat}&lon={lon}"))
             })
-            .or_else(|| self.config.city_id.as_ref().map(|x| format!("id={x}")))
-            .or_else(|| self.config.place.as_ref().map(|x| format!("q={x}")))
-            .error("no location was provided")?;
+            //Check cache, since place and zip lookups are expensive
+            .or_else(|| LOCATION_CACHE.lock().unwrap().clone());
+
+        location_query = match location_query {
+            Some(x) => Some(x),
+            None => match self.config.place.as_ref() {
+                Some(place) => {
+                    let url = format!("{GEO_URL}/direct?q={place}&appid={api_key}");
+                    REQWEST_CLIENT
+                        .get(url)
+                        .send()
+                        .await
+                        .error("Geo request failed")?
+                        .json::<Vec<CityCoord>>()
+                        .await
+                        .error("Geo failed to parse json")?
+                        .first()
+                        .and_then(|city| {
+                            let location = Some(format!("lat={}&lon={}", city.lat, city.lon));
+                            *LOCATION_CACHE.lock().unwrap() = location.clone();
+                            location
+                        })
+                }
+                None => None,
+            },
+        };
+
+        location_query = match location_query {
+            Some(x) => Some(x),
+            None => match self.config.zip.as_ref() {
+                Some(zip) => {
+                    let url = format!("{GEO_URL}/zip?zip={zip}&appid={api_key}");
+                    let city: CityCoord = REQWEST_CLIENT
+                        .get(url)
+                        .send()
+                        .await
+                        .error("Geo request failed")?
+                        .json()
+                        .await
+                        .error("Geo failed to parse json")?;
+
+                    let location = Some(format!("lat={}&lon={}", city.lat, city.lon));
+                    *LOCATION_CACHE.lock().unwrap() = location.clone();
+                    location
+                }
+                None => None,
+            },
+        };
+
+        location_query = match location_query {
+            Some(x) => Some(x),
+            None => match self.config.city_id.as_ref() {
+                Some(city_id) => {
+                    let city_id = &city_id.parse().error("city_id must be numeric")?;
+
+                    let mut cities_temp_path = env::temp_dir();
+                    let user = env::var("USER");
+                    cities_temp_path.push(format!(
+                        "city.list.min.json.gz-{}",
+                        user.as_deref().unwrap_or("no-user")
+                    ));
```

Comment:
Perhaps it's better to use `dirs::cache_dir()` instead of temp?

Reply:
By the way, maybe extract this cache and fetching logic into a different function?